spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -493,57 +635,15 @@ case class DeltaSharingSource(
       endOffset: DeltaSharingSourceOffset): DataFrame = {
     maybeGetFileChanges(startVersion, startIndex, isStartingVersion)
 
-    if (refreshPresignedUrls &&
-      (CachedTableManager.INSTANCE.preSignedUrlExpirationMs + lastQueryTableTimestamp -
-        System.currentTimeMillis() < CachedTableManager.INSTANCE.refreshThresholdMs)) {
-      // force a refresh if needed.
-      lastQueryTableTimestamp = System.currentTimeMillis()
-      val newIdToUrl = latestRefreshFunc()
-      sortedFetchedFiles = sortedFetchedFiles.map { indexedFile =>
-        IndexedFile(
-          version = indexedFile.version,
-          index = indexedFile.index,
-          add = if (indexedFile.add == null) {
-            null
-          } else {
-            val newUrl = newIdToUrl.getOrElse(
-              indexedFile.add.id,
-              throw new IllegalStateException(s"cannot find url for id ${indexedFile.add.id} " +
-                s"when refreshing table ${deltaLog.path}")
-            )
-            indexedFile.add.copy(url = newUrl)
-          },
-          remove = if (indexedFile.remove == null) {
-            null
-          } else {
-            val newUrl = newIdToUrl.getOrElse(
-              indexedFile.remove.id,
-              throw new IllegalStateException(s"cannot find url for id ${indexedFile.remove.id} " +
-                s"when refreshing table ${deltaLog.path}")
-            )
-            indexedFile.remove.copy(url = newUrl)
-          },
-          cdc = if (indexedFile.cdc == null) {
-            null
-          } else {
-            val newUrl = newIdToUrl.getOrElse(
-              indexedFile.cdc.id,
-              throw new IllegalStateException(s"cannot find url for id ${indexedFile.cdc.id} " +
-                s"when refreshing table ${deltaLog.path}")
-            )
-            indexedFile.cdc.copy(url = newUrl)
-          },
-          isLast = indexedFile.isLast
-        )
+    var fileActions = Seq[IndexedFile]()
+    synchronized {
+      fileActions = sortedFetchedFiles.takeWhile {
+        case IndexedFile(version, index, _, _, _, _) =>
+          version < endOffset.tableVersion ||
+            (version == endOffset.tableVersion && index <= endOffset.index)
       }
+      sortedFetchedFiles = sortedFetchedFiles.drop(fileActions.size)
     }
```

Comment:
Consider defining a method:
popSortedFetchFiles(endOffset)

The synchronization should be localized to setters/getters/modifiers

Reply:
added.